test/Libraries/RevitIntegrationTests/RegressionTests.cs

```diff
@@ -79,6 +82,155 @@ public void Regressions(RegressionTestData testData)
+        }
+
+        [Test]
+        [TestModel(@".\empty.rfa")]
+        public void NoDuplicatePortsOnUINodes()
+        {
+            // This test verifies Revit UI Nodes are not producing duplicate
+            // outports upon deserialization due to a lack of a JSON constructor
+
+            // Assertion variables
+            List<int> preSaveOutPortCount = new List<int>();
+            List<int> postSaveOutPortCount = new List<int>();
+            List<string> nodeCreationNames = new List<string>();
+            string outputData = "";
+
+            // Setup home ws
+            var homespace = Model.CurrentWorkspace as HomeWorkspaceModel;
+            Assert.NotNull(homespace, "The current workspace is not a HomeWorkspaceModel");
+
+            // Iterate through all loaded nodes in library & add Revit UI nodes to ws
+            var nodeList = Model.SearchModel.SearchEntries.OfType<NodeSearchElement>();
+            foreach (var node in nodeList)
+            {
+                var searchElement = node as NodeSearchElement;
+                var assembly = Path.GetFileName(node.Assembly);
+
+                if (assembly == "DSRevitNodesUI.dll")
+                {
+                    var currentNode = searchElement.CreateNode();
+                    Model.AddNodeToCurrentWorkspace(currentNode, true);
+                    nodeCreationNames.Add(searchElement.CreationName);
+                }
+            }
+
+            // Build list of outport counts for each active nodes before saving
+            foreach (var node in ViewModel.CurrentSpace.Nodes)
+            {
+                preSaveOutPortCount.Add(node.OutPorts.Count);
+            }
+
+            // Save workspace
+            ViewModel.CurrentSpace.Save(Path.Combine(workingDirectory, @".\DupePorts.dyn"));
+
+            // Close workspace
+            Assert.IsTrue(ViewModel.CloseHomeWorkspaceCommand.CanExecute(null));
+            ViewModel.CloseHomeWorkspaceCommand.Execute(null);
+
+            // Open Json temp file
+            OpenAndAssertNoDummyNodes(Path.Combine(workingDirectory, @".\DupePorts.dyn"));
+            RunCurrentModel();
+
+            // Build list of outport counts for each sucessfully restored nodes after saving/opening
+            foreach (var node in ViewModel.CurrentSpace.Nodes)
+            {
+                postSaveOutPortCount.Add(node.OutPorts.Count);
+            }
+
+            // PortCount lists have matching lengths pre/post save
+            Assert.IsTrue(preSaveOutPortCount.Count == postSaveOutPortCount.Count);
+            // PortCount values for each node match pre/post save
+            for (int i = 0; i < preSaveOutPortCount.Count; i++)
+            {
+                // Write node name to the xml results if assertion is false
+                if (preSaveOutPortCount[i] != postSaveOutPortCount[i])
+                {
+                    outputData += nodeCreationNames[i] + ", ";
+                }
```

Comment:
build a string of all the nodes with node port count discrepancies when comparing the pre-save state to the post-save state.  This is then written to the output xml when the test fails to aid debugging